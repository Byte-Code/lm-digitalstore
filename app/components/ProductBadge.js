import React from 'react';
import PropTypes from 'prop-types';
import ImmutablePropTypes from 'react-immutable-proptypes';
import { Map } from 'immutable';
import glamorous from 'glamorous';

import Image from './Image';
import { formatPrice, getProductAvailability } from '../utils/utils';
import MarketingFlag from '../components/MarketingFlag';
import { getMarketingProps } from '../utils/marketingUtils';

const ProductBadge = ({ productInfo, handleClick, animated, animatedDirection, stock }) => {
  if (productInfo.isEmpty()) {
    return null;
  }

  const imageID = productInfo.getIn(['basicInfo', 'data', 'mainImage']);
  const imageOptions = { width: 405, height: 405 };
  const name = productInfo.getIn(['basicInfo', 'data', 'name']);
  const grossPrice = productInfo.getIn(['price', 'data', 'selling', 'gross']);
  const listPrice = productInfo.getIn(['price', 'data', 'selling', 'list']);
  const isDiscounted = listPrice && listPrice - grossPrice > 1;
  const discount = productInfo.getIn(['price', 'data', 'selling', 'discount']);

  const MarketingFlagStyle = {
    wrapperStyle: {
      position: 'absolute',
      justifyContent: 'space-between !important',
      padding: '0px !important'
    },
    topLeftStyle: {
      '&>img': {
        marginLeft: '-5px !important',
        marginTop: '-4px'
      }
    },
    topRightStyle: {
      '&>img': {
        marginRight: '12px !important'
      }
    }
  };

  const marketingPriceProps = getMarketingProps(productInfo);

  return (
    <Wrapper animated={animated} animatedDirection={animatedDirection} onClick={handleClick}>
      <MarketingFlag {...marketingPriceProps} {...MarketingFlagStyle} />
      <Image imageID={imageID} imageOptions={imageOptions} alt={name} />
      <Name>{name}</Name>
      <PriceWrapper>
        {isDiscounted && <Discount>-{Math.round(discount)} %</Discount>}
        <Price isBarred={isDiscounted}>
          {formatPrice(listPrice) || formatPrice(grossPrice)} €
        </Price>
        {listPrice && <Price discounted>{formatPrice(grossPrice)} €</Price>}
      </PriceWrapper>
      <Available>
        <p>{getProductAvailability(productInfo, stock)}</p>
      </Available>
    </Wrapper>
  );
};

ProductBadge.propTypes = {
  productInfo: ImmutablePropTypes.map,
  handleClick: PropTypes.func,
  animated: PropTypes.bool,
  animatedDirection: PropTypes.string,
  stock: PropTypes.number
};

ProductBadge.defaultProps = {
  productInfo: Map(),
  handleClick: () => null,
  animated: false,
  animatedDirection: 'left',
  stock: 0
};

export default ProductBadge;

/* eslint-disable */
const Wrapper = glamorous.div(({ animated, animatedDirection }) => ({
  animation: animatedDirection === 'left'
    ? animated ?  'mymove 0.5s' : 'mymove1 0.5s'
    : animated ?  'mymoveRight 0.8s' : 'mymoveRight1 0.8s',
  height: 593,
  width: 405,
  display: 'flex',
  flexDirection: 'column',
  justifyContent: 'flex-start',
  position: 'relative',
  cursor: 'pointer',
  fontSize: 0,
  '&>div': {
    width: '100%',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',

  },
  '&>img': {
    height: 405,
    width: 405
  }
}));
/* eslint-enable */

const Name = glamorous.div({
  lineHeight: '32px',
  fontSize: 20,
  fontFamily: 'LeroyMerlinSans Bold',
  textTransform: 'uppercase',
  textAlign: 'center',
  margin: '10px 0'
});

const PriceWrapper = glamorous.div({
  display: 'flex',
  flexWrap: 'wrap',
  width: '100%',
  justifyContent: 'center',
  marginTop: '5%',
  '&>div:nth-child(3)': {
    marginLeft: 20
  }
});

const Price = glamorous.p(({ discounted, isBarred }) => ({
  fontSize: 20,
  lineHeight: '24px',
  color: discounted ? '#cc0000' : '#000',
  textDecoration: isBarred ? 'line-through' : 'none',
  marginLeft: '3%'
}));

export const Discount = glamorous.p({
  fontSize: 20,
  color: '#cc0000',
  textAlign: 'center',
  width: '100%'
});

export const Available = glamorous.div({
  color: '#339900',
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
  position: 'absolute',
  height: 40,
  bottom: 188,
  '&>p': {
    fontFamily: 'LeroyMerlinSans Bold',
    fontSize: 16
  },
  background: 'rgba(255, 255, 255, 0.8)'
});
