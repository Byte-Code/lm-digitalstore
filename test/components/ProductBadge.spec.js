import React from 'react';
import { shallow } from 'enzyme';
import { fromJS, Map } from 'immutable';

import ProductBadge, { Discount } from '../../app/components/ProductBadge';

const productInfo = fromJS({
  name: 'name',
  code: 'code',
  slug: 'slug',
  productDetail: {
    descriptionType: 'descriptionType',
    marketingDescriptions: {},
    descriptions: []
  },
  mainImage: 'mainImage',
  marketingAttributes: {},
  price: {
    selling: {
      list: 15,
      gross: 15
    }
  },
  storeStock: 10,
  pricingInformations: {},
  allStoreStock: [],
  images: ['image1'],
  loyaltyProgram: {},
  similarProducts: [
    {
      name: 'anche',
      products: ['prod1', 'prod2']
    }
  ]
});
const discountedPrice = fromJS({
  data: {
    selling: {
      list: 15,
      gross: 10,
      discount: 5
    }
  }
});
const productInfoDiscounted = productInfo.set('price', discountedPrice);
const handleClick = jest.fn();
const emptyProductInfo = Map();

describe('ProductBadge', () => {
  it('should return null when productInfo is empty', () => {
    const result = shallow(
      <ProductBadge
        handleClick={handleClick}
        productInfo={emptyProductInfo}
      />
    );
    expect(result).toMatchSnapshot();
  });

  it('should not display the discount if product is not discounted', () => {
    const result = shallow(
      <ProductBadge
        handleClick={handleClick}
        productInfo={productInfo}
      />
    );
    expect(result.find(Discount)).toHaveLength(0);
    expect(result).toMatchSnapshot();
  });

  it('should display the discount if product is discounted', () => {
    const result = shallow(
      <ProductBadge
        handleClick={handleClick}
        productInfo={productInfoDiscounted}
      />
    );
    expect(result.find(Discount)).toHaveLength(1);
    expect(result).toMatchSnapshot();
  });
});
