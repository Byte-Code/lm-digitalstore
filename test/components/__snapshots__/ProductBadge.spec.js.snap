// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductBadge should display the discount if product is discounted 1`] = `
<glamorous(div)
  animated={false}
  animatedDirection="left"
  onClick={[Function]}
>
  <MarketingFlag
    promotionCode="0"
    topLeftStyle={
      Object {
        "&>img": Object {
          "marginLeft": "-5px !important",
          "marginTop": "-4px",
        },
      }
    }
    topRightStyle={
      Object {
        "&>img": Object {
          "marginRight": "12px !important",
        },
      }
    }
    wrapperStyle={
      Object {
        "justifyContent": "space-between !important",
        "padding": "0px !important",
        "position": "absolute",
      }
    }
  />
  <Image
    fixBrightColor={false}
    imageOptions={
      Object {
        "height": 405,
        "width": 405,
      }
    }
    zoomable={false}
  />
  <glamorous(div) />
  <glamorous(div)>
    <glamorous(p)>
      -
      5
       %
    </glamorous(p)>
    <glamorous(div)
      isBarred={true}
    >
      15,00
       €
    </glamorous(div)>
    <glamorous(div)
      discounted={true}
    >
      10,00
       €
    </glamorous(div)>
  </glamorous(div)>
  <glamorous(div)>
    <p />
  </glamorous(div)>
</glamorous(div)>
`;

exports[`ProductBadge should not display the discount if product is not discounted 1`] = `
<glamorous(div)
  animated={false}
  animatedDirection="left"
  onClick={[Function]}
>
  <MarketingFlag
    promotionCode="0"
    topLeftStyle={
      Object {
        "&>img": Object {
          "marginLeft": "-5px !important",
          "marginTop": "-4px",
        },
      }
    }
    topRightStyle={
      Object {
        "&>img": Object {
          "marginRight": "12px !important",
        },
      }
    }
    wrapperStyle={
      Object {
        "justifyContent": "space-between !important",
        "padding": "0px !important",
        "position": "absolute",
      }
    }
  />
  <Image
    fixBrightColor={false}
    imageOptions={
      Object {
        "height": 405,
        "width": 405,
      }
    }
    zoomable={false}
  />
  <glamorous(div) />
  <glamorous(div)>
    <glamorous(div)>
       €
    </glamorous(div)>
  </glamorous(div)>
  <glamorous(div)>
    <p />
  </glamorous(div)>
</glamorous(div)>
`;

exports[`ProductBadge should return null when productInfo is empty 1`] = `null`;
